lib/seahorn/BvOpsem2WideMemMgr.cc

```diff
@@ -0,0 +1,543 @@
+                     m_ctx.alu().intTy(g_slotBitWidth));
+  }
+
+  PtrTy salloc(unsigned int bytes, uint32_t align) {
+    assert(isa<AllocaInst>(m_ctx.getCurrentInst()));
+    align = std::max(align, m_alignment);
+    auto region = m_allocator->salloc(bytes, align);
+    assert(region.first - region.second > 0);
+    return PtrTy(mkStackPtr(region.second).getRaw(),
+                 bytesToSlotExpr(region.first - region.second));
+  }
+
+  PtrTy salloc(Expr elmts, unsigned int typeSz, uint32_t align) {
+    align = std::max(align, m_alignment);
+
+    // -- compute number of bytes needed
+    Expr bytes = elmts;
+    if (typeSz > 1) {
+      // TODO: factor out multiplication and number creation
+      bytes = m_ctx.alu().doMul(bytes, m_ctx.alu().si(typeSz, ptrSzInBits()),
+                                ptrSzInBits());
+    }
+
+    // allocate
+    auto region = m_allocator->salloc(bytes, align);
+
+    // -- if allocation failed, return some pointer
+    if (m_allocator->isBadAddrInterval(region)) {
+      LOG("opsem", WARN << "imprecise handling of dynamically "
+                        << "sized stack allocation of " << *elmts
+                        << " elements of size" << typeSz << " bytes\n";);
+      return freshPtr();
+    }
+
+    // -- have a good region, return pointer to it
+    return PtrTy(
+        mkStackPtr(region.second).getRaw(),
+        m_ctx.alu().si(8 * (region.first - region.second), g_slotBitWidth));
+  }
+
+  PtrTy mkStackPtr(unsigned int offset) {
+    return PtrTy(m_main.mkStackPtr(offset), m_uninit_size);
+  }
+
+  PtrTy brk0Ptr() { return PtrTy(m_main.brk0Ptr(), m_uninit_size); }
+
+  PtrTy halloc(unsigned int _bytes, uint32_t align) {
+    PtrTy res = freshPtr();
+    LOG("opsem", WARN << "halloc() not implemented!\n");
+    return res;
+  }
+
+  PtrTy halloc(Expr bytes, uint32_t align) {
+    PtrTy res = freshPtr();
+    LOG("opsem", WARN << "halloc() not implemented!\n");
+    return res;
+  }
+
+  PtrTy galloc(const GlobalVariable &gv, uint32_t align) {
+    uint64_t gvSz = m_sem.getTD().getTypeAllocSize(gv.getValueType());
+    auto range = m_allocator->galloc(gv, gvSz, std::max(align, m_alignment));
+    return PtrTy(m_ctx.alu().si(range.first, ptrSzInBits()),
+                 bytesToSlotExpr(range.second - range.first));
+  }
+
+  PtrTy falloc(const Function &fn) {
+    auto range = m_allocator->falloc(fn, m_alignment);
+    return PtrTy(m_ctx.alu().si(range.first, ptrSzInBits()),
+                 bytesToSlotExpr(range.second - range.first));
+  }
+
+  // TODO: What is the right size to return here?
+  PtrTy getPtrToFunction(const Function &F) {
+    m_size.getPtrToFunction(F);
+    return PtrTy(m_main.getPtrToFunction(F), m_uninit_size);
+  }
+
+  PtrTy getPtrToGlobalVariable(const GlobalVariable &gv) {
+    uint64_t gvSz = m_sem.getTD().getTypeAllocSize(gv.getValueType());
+    return PtrTy(m_ctx.alu().si(
+                     m_allocator->getGlobalVariableAddr(gv, gvSz, m_alignment),
```

Comment:
a similar thing can be done for `getPtrToFunction` above.

Reply:
Let this be for now... added todo since this is not on critical path imo.